### PR TITLE
Some asm optimisation & update of the API.

### PR DESCRIPTION
**Pull request by ithinuel**
I adjusted the datastructures and managed shave-off a couple instructions from the asm impl.

### CODE REVIEW

File: examples/flash_example.rs [hunk 1]
```diff
@@ -48,8 +48,13 @@ impl FlashBlock {
         let addr = self.addr() - 0x10000000;
         defmt::assert!(addr & 0xfff == 0);
 
+        let flash_iap = flash::FlashIAP::new(true);
+
         cortex_m::interrupt::free(|_cs| {
-            flash::flash_range_erase_and_program(addr, data, true);
+            // SAFETY: we made sure the requirements are met
+            unsafe {
```
Comment (jannic) [inline examples/flash_example.rs]:
```
warning: unnecessary `unsafe` block
  --> examples/flash_example.rs:55:13
   |
47 |     unsafe fn write_flash(&self, data: &[u8; 4096]) {
   |     ----------------------------------------------- because it's nested under this `unsafe` fn
...
55 |             unsafe {
   |             ^^^^^^ unnecessary `unsafe` block
   |
   = note: `#[warn(unused_unsafe)]` on by default
   = note: this `unsafe` block does contain unsafe operations, but those are already allowed in an `unsafe fn`
   = note: `#[allow(unsafe_op_in_unsafe_fn)]` on by default

warning: `rp2040-flash` (example "flash_example") generated 1 warning
```

File: src/lib.rs [hunk 2]
```diff
@@ -1,241 +1,188 @@
...
-            flash_range_program: if write {
-                Some(rom_data::flash_range_program::ptr())
-            } else {
-                None
-            },
-            flash_flush_cache: rom_data::flash_flush_cache::ptr(),
-            flash_enter_cmd_xip: rom_data::flash_enter_cmd_xip::ptr(),
-            phantom: PhantomData,
-        }
+    pub struct FlashIAP {
```
Comment (jannic) [inline src/lib.rs:22]: What does IAP mean?

File: src/lib.rs [hunk 3]
```diff
@@ -1,241 +1,188 @@
...
+                    flash_enter_cmd_xip: rom_data::flash_enter_cmd_xip::ptr(),
+                },
+                boot2: [0u32; 256 / 4],
+            };
+            if use_boot2 {
+                unsafe {
+                    rom_data::memcpy44(&mut me.boot2 as *mut _, 0x10000000 as *const _, 256);
+                    let boot2_fn_ptr = (&me.boot2 as *const u32 as *const u8).offset(1);
+                    let boot2_fn: unsafe extern "C" fn() -> () = core::mem::transmute(boot2_fn_ptr);
+                    me.fn_ptrs.flash_enter_cmd_xip = boot2_fn;
```
Comment (jannic) [inline src/lib.rs]: I don't think this is correct: This essentially creates a pointer into `Self`, but `Self` can move at any time. I guess that this already breaks if you disable optimizations, but even with optimizations, it's not guaranteed that the struct is created in-place and not copied to the caller.

**Review by jannic — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 2 above)
Comment (ithinuel) [reply src/lib.rs:22]: `In-Application Programming`

It's what's commonly (in my experience) used to refer to drivers that write to the same storage as where the executable is running from.

EDIT: There is a dash between `In` and `Application`

File: src/lib.rs (hunk 3 above)
Comment (ithinuel) [reply src/lib.rs]: hm that's true, I did'nt think about that.
This resolution (lines 43-45) can be moved to `write_flash_inner` so that it's computed right before jumping to ram.

EDIT: Done

File: examples/flash_example.rs (hunk 1 above)
Comment (ithinuel) [reply examples/flash_example.rs]: Interesting, nightly does not warn for that 🤔

Comment (jannic) [reply examples/flash_example.rs]: Probably because the long-term goal is to `#[deny(unsafe_op_in_unsafe_fn)]`, which means that the explicit `unsafe` block becomes necessary.

So the right fix might be to add `#![deny(unsafe_op_in_unsafe_fn)]` at the top of the file, and keep the `unsafe` block.

File: src/lib.rs [hunk 4]
```diff
@@ -104,22 +106,29 @@ pub mod flash {
...
+                self.fn_ptrs.flash_enter_cmd_xip = boot2_fn;
+            }
             self.write_flash_inner_in_ram(addr, data_ptr, len, erase)
         }
 
         #[inline(never)]
         #[link_section = ".data.ram_func"]
         unsafe fn write_flash_inner_in_ram(
-            &self,
+            &mut self,
```
Comment (jannic) [inline src/lib.rs:131]: Why did you add `mut`, here?

Comment (ithinuel) [reply src/lib.rs:131]: Mostly to be consistent with the mur on the other method.

The erase/write modify the flash memory that's represented (although the compiler does not know that)
